### ISSUE TITLE
Allow to not automatically install software-properties-common

### ISSUE THREAD

**Issue by serbrech** [juju/juju#10593]
I use the cloudinit package to build the our scripts and cloud-init configs.
I would like to be able to control the versions of all packages being install on the machines.  
Unfortunately, the `software-properties-common` package is installed without any way to opt-out or to provide a specific version.
I would like to be able to use juju's `AddPackageSource` and to control the installation of the software-properties-common version.

https://github.com/juju/juju/blob/1a4d691a93084d91933e4abfcfa7650b4cdc9afa/cloudconfig/cloudinit/cloudinit_ubuntu.go#L161-L165

**Comment by mitechie** [comment]
Thanks, let me double check into that. I don't think Juju itself uses any ppa's or special apt repos any longer. Possible fallout might be to charms expecting the tooling to be there. We'll have to see. 

Normally, we'd suggest folks leverage an apt-proxy and controls on that end vs removing this as it's not actually protecting anyone from getting a different package during charms steps or the like. It's just the scripts for repo mgt vs editing the sources files directly.

**Comment by jameinel** [comment]
Hi @serbrech. I'd like to make sure I understand how you're wanting to use the code. It sounds like you're actually just wanting a cloud-init library for go, vs this library was sort of "what do we need to support enough of cloud-init for Juju's use case".

software-properties-common is an interesting issue, because we install that so we can get the 'add-apt-repository' command, so that we can specify what repositories you'll depend on.  (So if you wanted to install s-p-c from a PPA, there is a chicken-and-egg problem of not being able to add a repository without the command to add one.)

For *our* use case, we prefer not specifying a version, because we target many versions of Ubuntu (Trusty, Xenial, Bionic, Eoan, etc), and each of those likely has a different version.

If you wanted to provide a patch that lets you do something like OverridePackage("package-name", "package-name=version") or something along those lines, that would probably be acceptable.
I think here in Canonical we've also talked about having an 'official' cloud-init wrapper in Go, rather than our simplistic one.

**Comment by serbrech** [comment]
@jameinel that's correct. I use the cloud-init package directly to generate my own cloud-config or script to setup a host.
However, I want to be very specific about the version of each package, to make sure we do not change some dependencies automatically during the lifecycle of the host.
It seems that we start to reach the limit of what this cloud-init package is designed for. I'll evaluate your `OverridePackage` suggestion. thanks for getting back to me on the subject. @jameinel and @mitechie
